### DISCUSSION TITLE
Question about send behaviour

### DISCUSSION THREAD

**Discussion by schuelermine** [oakserver/oak#355] [Q&A]
I have a few questions about the behaviour of `send`.

What happens when `send` fails in various ways? Will the response object remain unwritable?  
Does `send` resolve `..` segments in paths? If so, will it ever backtrack beyond the root?  
Does `send` recognize trailing slashes and instead serve the index of the directory, or is the `index` parameter only for the root?

**Comment by somebody1234** [comment]
> What happens when `send` fails in various ways? Will the response object remain unwriteable?

not sure what you mean by it being unwriteable, given that `response` is never `Readonly` afaict.

you should be able to just wrap in a `try`/`catch` and handle failure cases in the `catch` block.
https://github.com/oakserver/oak/blob/4de406aaf221bbf6a856f99dc9d9cd10ed8de4b3/send.ts#L243-L246

> Does `send` resolve `..` segments in paths? If so, will it ever backtrack beyond the root?

yes to the first, no to the second:
https://github.com/oakserver/oak/blob/4de406aaf221bbf6a856f99dc9d9cd10ed8de4b3/send.ts#L195

https://github.com/oakserver/oak/blob/4de406aaf221bbf6a856f99dc9d9cd10ed8de4b3/util.ts#L316-L319

> Does `send` recognize trailing slashes and instead serve the index of the directory, or is the `index` parameter only for the root?

the index parameter is a default filename to serve for directories, usually `index.html`:
https://github.com/oakserver/oak/blob/4de406aaf221bbf6a856f99dc9d9cd10ed8de4b3/send.ts#L234-L236

**Reply by schuelermine** [reply]
AFAIK writing to `context.request.body` makes it readonly, even if the write fails.

```ts
app.use(async (context, next) => {
    try {
        context.request.body = (() => {throw null})() as any;
    } catch {
        await next();
    }
})

app.use(context => {
    context.response.body = ""
})
```
↑ this will fail

**Reply by schuelermine** [reply]
wait no

**Reply by schuelermine** [reply]
that seems to work, fine
I don’t know why it didn’t work in another situation where I tried something similar
